### PR TITLE
[pipelining] fix py ref cycle in stage_backward

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136507

TLDR; found forward activation tensors were being kept alive "forever"
(or until GC ran), and tracked it down to a cycle involving
`stage_backward.<locals>.extract_tensors_with_grads`.

The reference cycle in question is below.  (constructed using gc.get_referrers after doing a gc.collect in gc debug mode)

tensor is kept alive by
`[(<class 'cell'>, '0x7f7360234400')]`

tuple of cell objects
`(<cell at 0x7f73602343d0: function object at 0x7f734fff0ee0>, <cell at 0x7f7360234400: list object at 0x7f734e4d9a80>, <cell at 0x7f73602a4190: list object at 0x7f734eff8b00>)`
is kept alive by
`[(<class 'function'>, '0x7f734fff0ee0')]`

`<function stage_backward.<locals>.extract_tensors_with_grads at 0x7f734fff0ee0>`
is kept alive by 
`[(<class 'cell'>, '0x7f73602343d0')]`

Put into more plain terms,

```

def stage_backward(...):
    ...
    stage_output_tensors = []

    # a cell object will exist that contains the variables defined in stage_backward and used by
    # both stage_backward and nested functions
    # in this case, the cell object contains 'stage_output_tensors' but 

    # this function object will hold a reference to a 'cell' that contains any vars from
    # the parent scope not explicitly passed into the function as args.
    def extract_tensors_with_grads(...):
        ...
            # extract_tensors_with_grads refers to stage_output_tensors, so stage_output_tensors
            # is in the cell
            stage_output_tensors.append(output_val)
        ...
            # but extract_tensors_with_grads ALSO refers to itself (extract_tensors_with_grads),
            # so `extract_tensors_with_grads` will be in the cell
            extract_tensors_with_grads(...)
```

More debug details:
https://docs.google.com/document/d/1QPH1Lz0tnieIFPM2tyHrjVB-bjlnHuDgjx1p2am3cmE/edit?usp=sharing

In pdb:
```
gc.collect()
g = gc.garbage
g[-1]
[rank0]:(Pdb) [rank0]:<function
stage_backward.<locals>.extract_tensors_with_grads at 0x7fee5c3392d0>
g[-2]
[rank0]:(Pdb) [rank0]:(<cell at 0x7fee7abbcf40: function object at
0x7fee5c3392d0>, <cell at 0x7fee7abbcf70: list object at
0x7fee7ab68940>, <cell at 0x7fee5c3210c0: list object at 0x7fee5e1
d6340>)
g[-3]
[rank0]:(Pdb) [rank0]:[tensor([[[-4.1127e-06, -3.3826e-06,  2.6226e-06,
...,  6.4969e-06,
[rank0]:          -4.4405e-06, -4.7684e-06],
```

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

**Review by awgu — APPROVED** [review]

File: torch/distributed/pipelining/_backward.py [hunk 1]
```diff
@@ -320,6 +320,8 @@ def extract_tensors_with_grads(output_val, grad_val):
         torch.autograd.backward(
             stage_output_tensors, grad_tensors=output_grad_tensors  # type: ignore[arg-type]
         )
+        del stage_output_tensors
+        del output_grad_tensors
```
Comment (kwen2501) [inline torch/distributed/pipelining/_backward.py:324]: Interesting, is this related to how `stage_backward` is called? i.e. lambda function
https://github.com/pytorch/pytorch/blob/3be150653cedb51f4588b15a46dc0522de672f9f/torch/distributed/pipelining/stage.py#L490-L495

**Review by kwen2501 — APPROVED** [review]

File: torch/distributed/pipelining/_backward.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/pipelining/_backward.py:324]: no, this is unrelated.
I'm adding additional information to the doc and PR-desc

File: torch/distributed/pipelining/_backward.py [hunk 2]
```diff
@@ -320,6 +320,8 @@ def extract_tensors_with_grads(output_val, grad_val):
         torch.autograd.backward(
             stage_output_tensors, grad_tensors=output_grad_tensors  # type: ignore[arg-type]
         )
+        del stage_output_tensors
```
Comment (wconstab) [inline torch/distributed/pipelining/_backward.py:323]: after digging deeper into what the ref cycle was, I prefer to change this to `del extract_tensors_with_grads` which breaks the cycle and ends up freeing the tensors as well as other pyobjs that would have remained in a cycle if i only `del stage_output_tensors` and `del output_grad_tensors`

Comment (wconstab) [reply torch/distributed/pipelining/_backward.py]: i've updated the PR with a fix that is unfortunately a bit uglier, but arguably more correct.  If folks have a better idea, let me know

Comment (wconstab) [reply torch/distributed/pipelining/_backward.py]: ok- one more update: i don't really need to pass `stage_output_tensors` and `output_grad_tensors` into `extract_tensors_with_grads`.  I'll let those be captured in the cell.  I just need to ensure the cell will die, and for that i just have to break the cycle caused by the self-reference.  And that's a little less ugly.

**Review by kwen2501 — APPROVED** [review]
Thanks for digging further! Wow, what a new learning!

File: torch/distributed/pipelining/_backward.py [hunk 3]
```diff
@@ -303,19 +308,35 @@ def extract_tensors_with_grads(output_val, grad_val):
...
-        extract_tensors_with_grads(stage_output, output_grads)
+        # Note: ref cycle
+        # break a ref cycle that would keep tensors alive until GC runs
+        # 1. extract_tensors_with_grads refers to a cell that holds refs to any vars defined in stage_backward
+        #    and used in extract_tensors_with_grads
+        # 2. extract_tensors_with_grads referred to both stage_output_tensors, output_grad_tensors,
+        #    and to itself (extract_tensors_with_grads) since it makes a recursive call
+        # 3. stage_output_tensors was kept alive by the above refcycle, and it holds activation tensors, which is bad
+        # fix -> explictly pass in the ref to the fn, so there is no gc cycle anymore
+        extract_tensors_with_grads(
```
Comment (H-Huang) [inline torch/distributed/pipelining/_backward.py:337]: if you passed in `stage_output_tensors` and `output_grad_tensors` without the `extract_tensors_with_grads` ref, would that also fix the reference cycle? Since then `extract_tensors_with_grads` will no longer have to access variables in the outer scope so there should be no cycle, or is my understanding off.

Comment (wconstab) [reply torch/distributed/pipelining/_backward.py:337]: no, that would fix the tensor lifetime problem, which is almost as good.  However technically, there would still be a reference cycle between `extract_tensors_with_grads` and the cell.  This cycle would live on until GC kicked in.  So breaking the cycle is both cleaner (less vars passed in than passing in the two lists), and better (not only prevents leaking the tensors, but also cleans up the pyobjs earlier)

Comment (kwen2501) [reply torch/distributed/pipelining/_backward.py:337]: Following up: can we move `extract_tensors_with_grads` out of the scope of `stage_backward`?

Comment (wconstab) [reply torch/distributed/pipelining/_backward.py:337]: yes, but, then we'd be forced to add the other 2 lists back as args to extract_tensors_with_grads, but we could delete the `extract_tensors_with_grads` argument since `extract_tensors_with_grads` is a global now instead of a cellvar.  If you guys want to make that change, its fine too.
